spec/compiler/crystal/tools/unreachable_spec.cr

```diff
@@ -0,0 +1,342 @@
+  end
+
+  it "finds instance methods in nested types" do
+    assert_unreachable <<-CR
+      module Mod
+        class Foo
+          ༓def foo
+            1
+          end
+
+          def bar
+            2
+          end
+        end
+      end
+
+      Mod::Foo.new.bar
+      CR
+  end
+
+  it "finds method with free variable" do
+    assert_unreachable <<-CR
+      ༓def foo(u : U) forall U
+      end
+
+      def bar(u : U) forall U
+      end
+
+      bar(1)
+      CR
+  end
+
+  it "finds yielding methods" do
+    assert_unreachable <<-CR
+      ༓def foo
+        yield
+      end
+
+      def bar
+        yield
+      end
+
+      bar {}
+      CR
+  end
+
+  it "finds method called from block" do
+    assert_unreachable <<-CR
+      ༓def foo
+      end
+
+      def bar
+      end
+
+      def baz
+        yield
+      end
+
+      baz do
+        bar
+      end
+      CR
+  end
+
+  it "finds method called from proc" do
+    assert_unreachable <<-CR
+      ༓def foo
+      end
+
+      def bar
+      end
+
+      def baz(&proc : ->)
+        proc.call
+      end
+
+      baz do
+        bar
+      end
+      CR
+  end
+
+  it "finds methods with proc parameter" do
+    assert_unreachable <<-CR
+      ༓def foo(&proc : ->)
+        proc.call
+      end
+
+      def bar(&proc : ->)
+        proc.call
+      end
+
+      bar {}
+      CR
+  end
+
+  it "finds shadowed method (1)" do
+    assert_unreachable <<-CR
+      ༓def foo
+      end
+
+      ༓def foo
+      end
+
+      ༓def bar
+      end
+
+      def bar
+      end
+
+      bar
+      CR
+  end
+
+  it "finds method with `previous_def`" do
+    assert_unreachable <<-CR
+      ༓def foo
+      end
+
+      ༓def foo
+        previous_def
+      end
+
+      def bar
+      end
+
+      def bar
+        previous_def
+      end
+
+      bar
+      CR
+  end
+
+  it "finds methods called from reachable code" do
+    assert_unreachable <<-CR
+      ༓def qux_foo
+      end
+
+      ༓def foo
+        qux_foo
+      end
+
+      def qux_bar
+      end
+
+      def bar
+        qux_bar
+      end
+
+      bar
+      CR
+  end
+
+  it "does not finds method with `super`" do
+    assert_unreachable <<-CR
+      class Foo
+        ༓def foo
+        end
+
+        def bar
+        end
+      end
+
+      class Qux < Foo
+        ༓def foo
+          super
+        end
+
+        def bar
+          super
+        end
+      end
+
+      Qux.new.bar
+      CR
+  end
+
+  it "finds methods in generic type" do
+    assert_unreachable <<-CR
+      class Foo(T)
+        ༓def foo
+          1
+        end
+
+        def bar
+          2
+        end
+      end
+
+      Foo(Int32).new.bar
+      CR
+  end
+
+  it "finds method in abstract type" do
+    assert_unreachable <<-CR
+      abstract class Foo
+        ༓def foo
+        end
+
+        def bar
+        end
+      end
+
+      class Baz < Foo
+      end
+
+      Baz.new.bar
+      CR
+  end
+
+  # TODO: Should abstract Foo#bar be reported as well?
```

Comment:
I believe so. Otherwise we might never find that the method is no longer needed. But I would understand if not reported.

If the entry point is specs those might mark it as used, but without specs then it could be pointed out.

If only the abstract is there when building the source, should we get an abstract check error?

Reply:
> If only the abstract is there when building the source, should we get an abstract check error?

Yes, that's how it works currently. If an abstract method is never called, the tool reports all implementations as unreachable. But not the abstract def. Then after removing all the reported implementations, you'll get an abstract check error.

I think that's good enough for an MVP implementation.